### PR TITLE
Add Typescript type declarations

### PR DESCRIPTION
**Pull request by ianks**
This is an initial stab at creating some type declarations for this library. It basically just takes the jsdoc annotations, and converts this to types. While this is not perfect, it is a good place to start and I can refine the types over time. 

Let me know know what you think; it would make this library much more accessible to Typescript users and add value to this already awesome library!

Cheers,
Ian

### CODE REVIEW

**Comment by tommikaikkonen** [comment]
Sorry for getting back to this so late -- this is very appreciated!

I'll be merging this in for the next release as it's just type information and doesn't affect functionality; However note that I do not use Typescript myself and this adds duplication of the docs and class interfaces, so it's possible that me or someone else forgets to apply their changes to the type declarations in future development. For those cases just open up an issue or a PR :) Most of the API is pretty stable, so this shouldn't be a big problem.

**Comment by Aryk** [comment]
Hey @tommikaikkonen, when do you think the next release will be?

**Comment by ianks** [comment]
Also, please contribute to make the typing material better for anyone who uses this! There are currently many any types and such that can be better specified :)

**Comment by tommikaikkonen** [comment]
Unfortunately the next release (#62) has some pretty big changes, which postpone this. @ianks what was your process for compiling the type definitions, did you write them manually based on the docstrings? On the good side the API is smaller with #62 . Once the 0.9 moves from release candidate to proper release I'll look into updating this PR for that.

Note that if you want to use these type definitions right away with 0.8, you can fork the project from this branch from @ianks [and follow instructions here](https://github.com/tommikaikkonen/redux-orm/issues/63) to be able to point to the fork from `package.json`.

**Comment by Aryk** [comment]
Yeah, I'm actually using them and have since added to it, just got your new release candidate 0.9.0-rc.0, will continue to modify TS files.

**Comment by Aryk** [comment]
Here is my updated typings so far for anyone interested:

https://gist.github.com/Aryk/860725e2306c5ef485e528d51b536efd

**Comment by markerikson** [comment]
On a similar note: has anyone actually tried to define Flow types yet?  @Aryk , I tried running your TS typing through the `flowgen` tool, but it's clearly failing to convert some things properly.

**Comment by ahutchings** [comment]
@markerikson I'm using one based on @Aryk's typescript typing. I plan to contribute it to flow-typed once it's been refined a bit more.

https://gist.github.com/ahutchings/357001da662addb30e45e15fe87206e2

**Comment by markerikson** [comment]
@ahutchings : sweet, thanks!  I'll try that out shortly.

Two thoughts:

- I think both the TS and Flow typings are a bit wrong in that `Model.modelName` should be static, and I don't see a `static` keyword in front of that (ditto for a couple of the other fields?)
- Do you have any examples of how to actually _use_ those typings in terms of the rest of the app?  For example, if I have `const theModel = session.SomeModel.withId(123)`:
  - How are you defining types for your own model classes, particularly in regards to having the inherited fields from `Model` and the generated fields from the data and the relations?
  - How are you importing and using those types in relation to retrieving values from the session, especially since the session -specific classes are derived at runtime?

Basically, I'd love to see a quick example of actually using Redux-ORM with Flow.  If there's any chance you could post a few snippets or even point to a repo, I'd really appreciate it!

**Comment by ahutchings** [comment]
@markerikson I found and fixed a few fields that were not marked as `static` in the typings, but haven't done a full review to ensure I caught them all. I will update to fix the `modelName` field you mentioned.

Regarding using the typings, at the moment I'm subclassing the ORM and Session classes and typing each model as instance fields on it like so:

```
// @flow
import {Session} from 'redux-orm';
import type Workspace from './domain/Workspace';
import type Console from './domain/Console';
import type Pane from './domain/Pane';
import type Region from './domain/Region';
import type Resizer from './domain/Resizer';
import type Widget from './domain/Widget';
import type Overlay from './domain/Overlay';
import type {Entities} from './types';

class WorkspacesSession extends Session {
  Console: Class<Console>;
  Overlay: Class<Overlay>;
  Pane: Class<Pane>;
  Region: Class<Region>;
  Resizer: Class<Resizer>;
  Widget: Class<Widget>;
  Workspace: Class<Workspace>;

  state: Entities;
}

class WorkspacesORM extends ORM {
  getEmptyState(): Entities {
    return super.getEmptyState();
  }

  emptySession() {
    const state = this.getEmptyState();
    return this.session(state);
  }

  session(state: Entities) {
    return new WorkspacesSession(this, this.getDatabase(), state);
  }

  mutableSession(state: Entities) {
    return new WorkspacesSession(this, this.getDatabase(), state, true);
  }
}
```

I think it would be better to use parameterized types to "inject" the domain and state shapes into the base `ORM` and `Session` type definitions, but haven't put any time into that yet since subclassing is working fairly well.

**Comment by ianks** [comment]
I think at this point someone should submit the most up-to-date version to definitely typed and we should close this PR down.

**Comment by markerikson** [comment]
@ahutchings : thanks.  One more request: could you show an example those class types (such as `Widget` and `Workspace`), and that `Entities` type?

**Comment by haveyaseen** [comment]
Closing as per discussion in #168. Please use external repositories like @keenondrums' for TypeScript definitions.
